src/test/operators/join_test_runner.cpp

```diff
@@ -180,20 +197,92 @@ class JoinTestRunner : public BaseTestWithParam<JoinTestConfiguration> {
       std::make_shared<JoinOperatorFactory<JoinOperator>>(),
+      std::nullopt,
       std::nullopt
     };
     // clang-format on
 
-    const auto add_configuration_if_supported = [&](const auto& configuration) {
-      // String vs non-String comparisons are not supported in Hyrise and therefore cannot be tested
-      if ((configuration.data_type_left == DataType::String) != (configuration.data_type_right == DataType::String)) {
-        return;
+    /**
+     * Returns a set of adapted configurations if the join type provides further configuration possibilities that
+     * should be tested. Otherwise, a vector containing only the passed configuration is returned.
+     */
+    const auto build_join_type_specific_variations = [&](const auto& configuration) {
+      if constexpr (std::is_same_v<JoinOperator, JoinIndex>) {
+        // Currently the Index Join cannot deal with configurations where the join column
+        // data types are different, see Issue #2077
+        if (configuration.data_type_left != configuration.data_type_right) {
+          return std::vector<JoinTestConfiguration>{};
+        }
+
+        // For the JoinIndex, additional parameters influence its execution behavior:
+        //  - index side (left or right)
+        //  - availability of an index for a join column segment
+        //  - presence of "single chunk reference guarantee" for join column reference segments
+
+        // share of indexed segments; 1 means all segments are indexed
+        std::array<float, 2> indexed_segment_shares{1.0f, .1f};
+
+        std::vector<JoinTestConfiguration> variations{};
+        variations.reserve(all_index_sides.size() * indexed_segment_shares.size());
+        for (const auto& index_side : all_index_sides) {
+          // calculate index chunk counts, eliminate duplicates by using the unordered set
+          auto& indexed_input = index_side == IndexSide::Left ? configuration.input_left : configuration.input_right;
+          const auto chunk_count =
+              static_cast<uint32_t>(std::ceil(static_cast<float>(indexed_input.table_size) / indexed_input.chunk_size));
+
+          auto indexed_chunk_counts = std::unordered_set<uint32_t>{};
+          for (const auto indexed_share : indexed_segment_shares) {
+            indexed_chunk_counts.insert(static_cast<uint32_t>(indexed_share * chunk_count));
+          }
+
+          for (const auto indexed_chunk_count : indexed_chunk_counts) {
+            auto variation = configuration;
+            auto& variation_indexed_input =
+                index_side == IndexSide::Left ? variation.input_left : variation.input_right;
+            variation.index_side = index_side;
+            variation_indexed_input.indexed_chunk_count = indexed_chunk_count;
+
+            if (variation_indexed_input.table_type != InputTableType::Data && variation.join_mode == JoinMode::Inner) {
+              variation_indexed_input.single_chunk_reference_count = indexed_chunk_count;
```

Comment:
Do we want to guarantee this for each chunk?

Reply:
The idea was to have this guarantee for each indexed chunk and for an additional (non-indexed) one if the number of non-indexed segments is greater than 1. Therefore we cover (a) and (b) of the following execution branches:
  - (a) ref. segment with single chunk reference guarantee, indexed
  - (b) ref. segment with single chunk reference guarantee, not indexed (fallback nested loop)
  - (c) ref. segment with no single chunk reference guarantee (fallback nested loop)

But yes, (c) is not covered which might be your concern here? How about defining a range for segments with this guarantee that starts at a position larger than 0 and ends at a position larger than the last indexed segment? Let's say the first segment never provides this guarantee but all other indexed segments and an additional non-indexed do?

Example:
- segment 0, no guarantee, indexed --> covers (c)
- segment 1, guarantee, indexed --> covers (a)
- segment 2, guarantee, not indexed --> covers (b)
- segment 3, no guarantee, not indexed --> covers (c)

Reply:
Sounds good. :)